### PR TITLE
Preliminary pull request which converts a dataframe source to a ga4ghfeature

### PR DESCRIPTION
**Pull request by abhim00**

### CODE REVIEW

File: mango-viz/bdgenomics/mango/pileup/sources.py [hunk 1]
```diff
@@ -79,6 +80,20 @@ def __init__(self, json):
         #: dictionary containing source elements (viz, source, sourceOptions, label)
         self.dict_ = json
 
+class DataFrameSource(Source):
+    
+    def __init__(self, dataframe):
+        """Initializes modin dataframe. Converts it to raw json.
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/sources.py]: Should work with modin or pandas

File: mango-viz/bdgenomics/mango/pileup/utils.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import json
+
+def parse_bed(json_input):
+    """ Converts a bed JSON string to a GA4GHFeature JSON String
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/utils.py]: Would it maybe be easier to just pass the dataframe into this function and parse the rows directly? Then you wouldn't have to do all of the string replace work, and simply convert each row. You could use the function `build_json_from_bed` to map each row in the dataframe

File: mango-viz/bdgenomics/mango/pileup/utils.py [hunk 3]
```diff
@@ -1,74 +1,50 @@
 import json
+def parse_bed_dataframe(dataframe):
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/utils.py]: Looks good! Can you add unit tests for this?

Comment (abhim00) [reply mango-viz/bdgenomics/mango/pileup/utils.py]: where should I add these unit tests? also, is there an example of one?

Comment (akmorrow13) [reply mango-viz/bdgenomics/mango/pileup/utils.py]: Tests for mango-viz are at https://github.com/bigdatagenomics/mango/tree/master/mango-viz/bdgenomics/mango/pileup/test. You can add a test-utils.py file with tests

Comment (akmorrow13) [reply mango-viz/bdgenomics/mango/pileup/utils.py]: remove this file I think it is duplicated.

File: mango-viz/bdgenomics/io/utils.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    "strand", "thickStart", "thickEnd", "itemRGB", "blockCount", 
+    "blockSizes", "blockStarts"],
+    skiprows = None,
+):
+
+    column_names = ["chrom","chromStart", "chromEnd", "name", "score", 
+    "strand", "thickStart", "thickEnd", "itemRGB", "blockCount", 
+    "blockSizes", "blockStarts"]
+
+    return pandas.read_csv(filepath_or_buffer, sep = "\t", names = column_names, skiprows = skiprows)
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/io/utils.py]: add an option that checks if the file is compressed (.gz)

File: mango-viz/bdgenomics/io/utils.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import json
+import pandas
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/io/utils.py]: - In https://github.com/bigdatagenomics/mango/blob/master/mango-viz/requirements.txt put pandas and modin as requirements
- required functions for each file type (bed, vcf, etc): `read` (seen by the user), `_parseDataframe` (internal),`_toJson` (internal)


Usage:
```
df = bdgenomics.mango.io.read_bed(file, use_pandas = False) # use_pandas will be set by user
tr = Track(source = DataframeSource(df), ...)
```

File: mango-viz/bdgenomics/mango/io/bedfile.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                bed_content = "\"referenceName\":{}, \"start\":{}, \"end\":{}".format("\""+chromosomes[i]+"\"", "\""+str(chrom_starts[i])+"\"", "\""+str(chrom_ends[i])+"\"")
+                json_ga4gh = json_ga4gh + "{" + bed_content + "},"
+            else:
+                json_ga4gh = json_ga4gh[:len(json_ga4gh)-1]
+
+
+        #ending json
+        json_ga4gh = json_ga4gh + "]}"
+        return json_ga4gh
+        #raise NotImplementedError("Implement Me!")
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/bedfile.py]: remove this

File: mango-viz/bdgenomics/mango/io/genomicfile.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import modin.pandas as pd
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/genomicfile.py]: add headers to all of the files (you can reference the other py files as an example)

File: mango-viz/bdgenomics/mango/pileup/sources.py [hunk 8]
```diff
@@ -79,6 +80,20 @@ def __init__(self, json):
...
 
+class DataFrameSource(Source):
+    
+    def __init__(self, dataframe):
+        """Initializes dataframe. Converts it to raw json. 
+
+        Args:
+            :param dataframe: dataframe
+        """
+        feature_transformed_json = parse_bed_dataframe(dataframe)
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/sources.py]: I think the way @devin-petersohn set this up is that you should be able to call the _to_json function directly from the dataframe

File: mango-viz/bdgenomics/mango/pileup/sources.py [hunk 9]
```diff
@@ -79,6 +80,20 @@ def __init__(self, json):
...
+        self.dict_ = feature_transformed_json
+        
+    #: name that pileup.js uses to identify sources --> temporary
+    name = 'featureJson'
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/sources.py]: Is this being used anywhere?

File: mango-viz/bdgenomics/mango/pileup/sources.py [hunk 10]
```diff
@@ -79,6 +80,20 @@ def __init__(self, json):
...
 
+class DataFrameSource(Source):
+    
+    def __init__(self, dataframe):
+        """Initializes dataframe. Converts it to raw json. 
+
+        Args:
+            :param dataframe: dataframe
+        """
+        feature_transformed_json = read_bed(dataframe)._mango_to_json
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/sources.py]: don't call read_bed here

File: mango-viz/bdgenomics/mango/io/bedfile.py [hunk 11]
```diff
@@ -47,4 +46,7 @@ def _to_json(cls, df):
         #ending json
         json_ga4gh = json_ga4gh + "]}"
         return json_ga4gh
-        #raise NotImplementedError("Implement Me!")
+    
+    @classmethod
+    def _visualization(cls, df):
+        return 'featureJson'
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/bedfile.py]: We have been talking about adding unit tests for a while, can you add some to this PR?

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        chromosomes = list(dataframe["chrom"])
+        chromStart = list(dataframe["chromStart"])
+        chromEnd = list(dataframe["chromEnd"])
+        for item in chromsomes:
+            assert(type(item) == str)
+        for item in chromStart:
+            assert(type(item) == int)
+        for item in chromEnd:
+            assert(type(item) == int)
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: Some other unit test ideas include:

def test_from_dataframe(): make sure  a user can take a dataframe and convert it to a bed dataframe

You can also test:
- that the number or rows are correct
- that the json strings are correct

Comment (abhim00) [reply mango-viz/bdgenomics/mango/io/test/test_bed.py]: Currently, the read_bed functionality requires the input to be a path destination to a .bed file. User's cannot take an existing dataframe and convert it to a bed dataframe. Is that functionality that we want added ?

Comment (akmorrow13) [reply mango-viz/bdgenomics/mango/io/test/test_bed.py]: I believe this functionality already exists by using `def _parse(cls, df)` in mango-viz/bdgenomics/mango/io/bedfile.py. This should be testable.

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            json_object = json.loads(string)
+        except ValueError, e:
+            return False
+        return True
+
+    def test_to_json():
+        assert(type(dataframe._mango_to_json) == str)
+        assert(is_valid_json(dataframe._mango_to_json) == True)
+    
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: remove these lines

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+class BedFileTest(unittest.TestCase):
+
+    dataframe = io.read_bed("../../../examples/data/chr17.582500-594500.bed")
+    filename = "../../../examples/data/chr17.582500-594500.bed"
+
+    def test_required_columns():
+        dataframe_columns = list(dataframe.columns)
+        for name in ("chrom", "chromStart", "chromEnd"):
+            assert(name in dataframe_columns)
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: add another test for making a dataframe with ("chrom", "chromStart", "chromEnd") cols and going to a bed dataframe

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            assert(type(d1[i] == int))
+            assert(d1[i]==chromStart[i])
+
+        for i in range(len(chromEnd)):
+            assert(type(d2[i] == int))
+            assert(d2[i] == chromEnd[i])
+            
+        for i in range(len(chromosomes)):
+            assert(type(d3[i] == int))
+            assert(d3[i] == chromosomes[i])
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: Can you then pass this into the Visualization?

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+   
+
+
+
+    
+
+
+# Run tests
+if __name__ == '__main__':
+    unittest.main()
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: add newline

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        try:
+            json_object = json.loads(string)
+        except ValueError, e:
+            return False
+        return True
+
+    def test_to_json():
+        assert(type(dataframe._mango_to_json) == str)
+        assert(is_valid_json(dataframe._mango_to_json) == True)
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: Can you add tests to pass this into the Visualization? I dont think there currently are any tests to check whether the df works with the viz

File: mango-viz/bdgenomics/mango/pileup/track.py [hunk 18]
```diff
@@ -62,8 +62,10 @@ def __init__(self, **kwargs):
             :param kwargs: Should contain viz, optional source, optional label.
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/pileup/track.py]: remove changes

File: mango-viz/examples/data/chr17.582500-594500.bed [hunk 19]
```diff
@@ -0,0 +1,10 @@
+chr17	5821	31270
```
[the remark this answers is not in the document]
Comment (akmorrow13) [reply mango-viz/examples/data/chr17.582500-594500.bed]: bump @abhim00

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+        assert(str(type(reads)) == '<class \'bdgenomics.mango.pileup.pileupViewer.PileupViewer\'>')
+
+
+    
+        
+
+# Run tests
+if __name__ == '__main__':
+    unittest.main()
+
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: add newline

File: mango-viz/bdgenomics/mango/io/test/test_bed.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+            assert(type(d2[i] == int))
+            assert(d2[i] == chromEnd[i])
+
+        for i in range(len(chromosomes)):
+            assert(type(d3[i] == int))
+            assert(d3[i] == chromosomes[i])
+
+    
+    def test_validate_num_rows(self):
+        #dataframe = read_bed("../../../../examples/data/chr17.582500-594500.bed")
```
Comment (akmorrow13) [inline mango-viz/bdgenomics/mango/io/test/test_bed.py]: remove all unused code

**Comment by akmorrow13** [comment]
Jenkins, test this please.

**Comment by akmorrow13** [comment]
Jenkins, add to whitelist.

**Comment by akmorrow13** [comment]
Closed in favor of https://github.com/bigdatagenomics/mango/pull/527

**Pull request closed without merging**